internal/app/machined/pkg/controllers/network/dns_upstream.go

```diff
@@ -0,0 +1,162 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package network
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/coredns/coredns/plugin/pkg/proxy"
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/pkg/machinery/resources/config"
+	"github.com/siderolabs/talos/pkg/machinery/resources/network"
+)
+
+// DNSUpstreamController is a controller that manages DNS upstreams.
+type DNSUpstreamController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *DNSUpstreamController) Name() string {
+	return "network.DNSUpstreamController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *DNSUpstreamController) Inputs() []controller.Input {
```

Comment:
Maybe we case use `generic.Transform` controller here?

Reply:
There is 1->M relationship here. I actually asked @smira the same question, and he suggested that I do in plain and simple.